### ISSUE TITLE
Logout action does not terminate online status

### ISSUE THREAD

**Issue by wizgriz** [flarum/core#223]
1. Log in
2. Visit your profile, the status will say Online
3. Log out
4. Revisit your profile, it will still say Online

So I'm assuming it only expires at the timer. The status should be terminated immediately if a user manually logs out, no?

**Comment by Fastidious** [comment]
I have not seeing a forum that does this instantly. There is always a delay. Not saying this might not be a bug, but noting that what you describe is a prevalent behavior.

**Comment by wizgriz** [comment]
Then you haven't used xenforo, mybb, smf, phpbb, vbulletin or basically any forum software. I have no idea what you're talking about.

**Comment by Fastidious** [comment]
I have been a user in all those. The online "green" always lingers for a bit after the user have gone offline, even longer if the user never properly logged off (which almost no one does, on a frequently visited forum).

Again, I am not saying that it isn't a bug, but something I have seeing in general. For my own use, I could care less whether the user is online or not. I consider that an information not worth spending machine cycles over.

**Comment by wizgriz** [comment]
Lmao what you're saying is factually wrong and I have no idea where you're finding the reason to argue. This is a bug report like you say yourself and logical behavior is to set the correct state of a user when he logs out (which is in fact why you'll see this behavior in other forum software!)

**Comment by Fastidious** [comment]
I am not arguing. The problem with this type of interaction is that it might be source of misunderstandings. 

esoTalk behaves similarly to Flarum, around 5 minutes or so after user logoff it shows the user as logged off. Perhaps it is how @tobscure intended logout to work? Perhaps you have found a bug that has existed in esoTalk and now in Flarum? I am not dismissing either, just interacting.

**Comment by wizgriz** [comment]
Yea ok, esoTalk is "prevalent" forum software.

**Comment by s9e** [comment]
> I have been a user in all those. The online "green" always lingers for a bit after the user have gone offline

This isn't about going offline, this is about *logging out*. Clicking the Logout button, that is, which should immediately destroy the session and make a user appear offline.

**Comment by franzliedke** [comment]
Well, that's probably because Flarum is a single-page application. When you click on a link in Flarum, we don't load the whole new page from the server, but only the parts relevant to that page. Thus, things like the online status of other (!) users will likely not be updated very quickly.

Maybe we should make sure that that list will be refreshed at least every few minutes.

**Comment by s9e** [comment]
No, actually you can still see the user online in their profile even on another browser. So either the session isn't destroyed or the online indicator is cached.

**Comment by franzliedke** [comment]
Ah, okay; well that's a bug then, from my perspective ;)

**Comment by tobscure** [comment]
Not sure if this one is worth fixing, because it would add a bit of extra complexity:

* The user is said to be "online" if they were "seen" within the last 5 minutes. That last seen time is stored in the database.
* We can't simply set the last seen time to NULL upon logging out, because we don't want to lose that information.
* We would need to add another column, something like log_out_time. The user would then be "online" only if the log_out_time is less than the last_seen_time.

Thoughts?

**Comment by Fastidious** [comment]
It is such a minor thing, I do not think it is worth complicating things for it. If it was up to me, I will leave it as it is.

**Comment by wizgriz** [comment]
I think the extra column is a must. I rarely use keep-alive and if I do I never keep cookies alive for longer than the session. I usually manually log out of everything, and when I do, I expect the session to be entirely wiped.

Additionally, consider extensions like "who's online", or someone who would tweak the online grace period to be longer than 5 minutes--this information has now become sorely inaccurate and on display. I don't want to be seen online when I've specifically logged out, but I don't want to disable my online status entirely either.

**Comment by Luceos** [comment]
You can easily fix this using pusher. https://github.com/flarum/pusher/pull/2

Using one global presence channel you can now exactly (by the second) when someone is online or not. The only thing to do is listen to the necessary events of this channel. I gladly explain (or pr) the process in details if need by.

**Comment by tobscure** [comment]
@luceos That's handy, but the shortfalls of "idle timeout" are not what this discussion is about. This is about clearing the idle timeout when the Log Out button is pressed. :)

**Comment by Luceos** [comment]
Yes I see. I responded mainly on @wizgriz who wanted a "who's online" feature.

**Comment by tobscure** [comment]
@franzliedke and I have discussed this and think that we have two options:

* We can subtract 5 minutes from the last_seen_time when the user logs out – thus the user will no longer be considered "online", but some slightly incorrect information will be presented on their profile ("user was last active 5 minutes ago")
* OR we can just concede that a user will appear as "online" for a few minutes after they've logged out. Maybe we could even have a tooltip that says "Toby has been online in the past 5 minutes" or something.

We both agree though that the extra complexity of adding another column (log_out_time) is not worth it.

**Comment by tobscure** [comment]
@JoshyPHP, @wizgriz - what do you think?

**Comment by wizgriz** [comment]
If the complexity simply is not worth it, then I think this issue can be closed.

**Comment by franzliedke** [comment]
Then the only remaining question is: Do we want to update the tooltip as Toby suggested?

**Comment by franzliedke** [comment]
Oh, I would say yes.

**Comment by tobscure** [comment]
I agree. We should also make the 5 minute period configurable.

**Comment by franzliedke** [comment]
Configurable in the admin panel or via code (i.e. a config file)? 

I'd prefer the latter, so as not to clutter up the administration interface with rarely used options. If someone needs an UI for it, it can be added via an extension.

**Comment by tobscure** [comment]
Exactly. An entry in the `settings` table.

**Comment by buiductuan182** [comment]
This isn't about going offline, this is about logging out. Clicking the Logout button, that is, which should immediately destroy the session and make a user appear offline. 
https://www.quangminhhanoi.com/dieu-hoa-carrier

**Comment by clarkwinkelmann** [comment]
I wouldn't consider this a good first issue. It's unclear what exactly the issue is about at this point.

Should we close this and open separate issues? Re-reading this I see:

- Adding a label in the frontend to say that "online" means "seen in the last x minutes"
- Add a new setting in Flarum to configure how long the green disc shows up on users

Also discussed above is the use of a different method for online status, which I think we should delegate to community extensions. "seen in last x minutes" is good enough for core IMO.

**Comment by luceos** [comment]
We think it might be a great solution to remove the online features from core to flarum/pusher. Other extensions can provide a better online state than core can, especially websocket related extensions. The community can reintroduce similar behavior with a simple extension if they want to.

**Comment by clarkwinkelmann** [comment]
The "feature" that the online indicator is is really just a indicator component and a few lines of javascript that adds a green badge in place of the "last seen at" text on the profile by using the "last seen at" value as the source.

I still think the "dumb" indicator is useful in a discussion, it lets you know if the people that are interacting are still around.

I don't think we should remove it from core because (1) it's very little code and still useful for what it's worth and (2) it provides a component that community extension can use and that extensions providing real online status can replace.

I do think it would make sense for us to change the label text from "online" to "active recently" so people better understand what it means and can't complain about the lack of real time or expiration.

A real online indicator will definitely need some kind of websocket feature and would be great for a community extension, or possibly an option in our own Pusher extension.

To allow extensions to re-use the indicator we should maybe convert it into a component instead of the function we use now which isn't extensible https://github.com/flarum/core/blob/master/js/src/common/helpers/userOnline.tsx

Luckily we can do that in a backward-compatible way. Let's create a new component and return it with the helper function.
